syntax = "proto3";

package keto.proto;

import "google/protobuf/any.proto";

enum MessageType {
    MESSAGE_TYPE_TRANSACTION = 0;
    MESSAGE_TYPE_BLOCK = 1;
}

enum MessageOperation {
    MESSAGE_INIT = 0;
    MESSAGE_ROUTE = 1;
    MESSAGE_BALANCE = 2;
    MESSAGE_BLOCK = 3;
    MESSAGE_PROCESS = 4;
    MESSAGE_WAIT = 5;
    
    // Block Operations
    MESSAGE_PERSIST = 100;
}

/**
 * This is an example of a protobuf message.
 */
message MessageWrapper {
    
    uint32 version = 1;
    bytes session_hash = 2;
    bytes account_hash = 3;
    bytes source_account_hash = 4;
    MessageType message_type = 5;
    MessageOperation message_operation = 6;
    
    google.protobuf.Any msg = 7;
}

message MessageWrapperResponse {
    bool  success = 1;
    string result = 2;
}

message ForkMessageWrapper {
    string command = 1;
    string event = 2;
    bytes message = 3;
    string exception = 4;
}